src/couplings.jl

```diff
@@ -0,0 +1,244 @@
+    indices_3::AbstractVector{Int}
+)
+    A_1 = spzeros(n, length(indices_1));
+    A_2 = spzeros(n, length(indices_2));
+    A_3 = spzeros(n, length(indices_3));
+
+    for (i, idx) in enumerate(indices_1)
+        A_1[idx, i] = 1.0
+    end
+
+    for (i, idx) in enumerate(indices_2)
+        A_2[idx, i] = 1.0
+    end
+
+    for (i, idx) in enumerate(indices_3)
+        A_3[idx, i] = 1.0
+    end
+
+    return PartitionMask(A_1, A_2, A_3)
+end
+
+PartitionMask(
+    n::Int,
+    indices_1::AbstractVector{Int},
+    indices_2::AbstractVector{Int}
+) = PartitionMask(n, indices_1, indices_2, nothing)
+
+PartitionMask(
+    n::Int,
+    indices_1::AbstractVector{Int},
+    indices_2::AbstractVector{Int},
+    indices_3::Nothing
+) = PartitionMask(n, indices_1, indices_2, [i for i in 1:n if i ∉ (indices_1 ∪ indices_2)])
+
+PartitionMask(
+    n::Int,
+    indices_1::AbstractVector{Int},
+    indices_2::Nothing,
+    indices_3::AbstractVector{Int}
+) = PartitionMask(n, indices_1, [i for i in 1:n if i ∉ (indices_1 ∪ indices_3)], indices_3)
+
+"""
+    PartitionMask(n::Int, indices)
+
+Assumes you want to _split_ the vector, where `indices` refer to the 
+parts of the vector you want to apply the bijector to.
+"""
+function PartitionMask(n::Int, indices)
+    indices_2 = [i for i in 1:n if i ∉ indices]
+
+    # sparse arrays <3
+    A_1 = spzeros(n, length(indices));
+    A_2 = spzeros(n, length(indices_2));
+
+    # Like doing:
+    #    A[1, 1] = 1.0
+    #    A[3, 2] = 1.0
+    for (i, idx) in enumerate(indices)
+        A_1[idx, i] = 1.0
+    end
+
+    for (i, idx) in enumerate(indices_2)
+        A_2[idx, i] = 1.0
+    end
+
+    return PartitionMask(A_1, A_2, spzeros(n, 0))
+end
+PartitionMask(x::AbstractVector, indices) = PartitionMask(length(x), indices)
+
+"""
+    combine(m::PartitionMask, x_1, x_2, x_3)
+
+Combines `x_1`, `x_2`, and `x_3` into a single vector.
+"""
+@inline combine(m::PartitionMask, x_1, x_2, x_3) = m.A_1 * x_1 .+ m.A_2 * x_2 .+ m.A_3 * x_3
+
+"""
+    partition(m::PartitionMask, x)
+
+Partitions `x` into 3 disjoint subvectors.
+"""
+@inline partition(m::PartitionMask, x) = (transpose(m.A_1) * x, transpose(m.A_2) * x, transpose(m.A_3) * x)
+
+
+# CouplingLayer
+
+"""
+    CouplingLayer{B, M, F}(mask::M, θ::F)
+
+Implements a coupling-layer as defined in [1].
+
+# Examples
+```julia-repl
+julia> m = PartitionMask(3, [1], [2]) # <= going to use x[2] to parameterize transform of x[1]
+PartitionMask{SparseArrays.SparseMatrixCSC{Float64,Int64}}(
+  [1, 1]  =  1.0, 
+  [2, 1]  =  1.0, 
+  [3, 1]  =  1.0)
+
+julia> cl = CouplingLayer(Shift, m, identity) # <= will do `y[1:1] = x[1:1] + x[2:2]`;
+
+julia> x = [1., 2., 3.];
+
+julia> cl(x)
+3-element Array{Float64,1}:
+ 3.0
+ 2.0
+ 3.0
+```
+
+# References
+[1] Kobyzev, I., Prince, S., & Brubaker, M. A., Normalizing flows: introduction and ideas, CoRR, (),  (2019). 
+"""
+struct CouplingLayer{B, M, F} <: Bijector{1} where {B, M <: PartitionMask, F}
```

Comment:
`B` can be a field. Type parameters must be `isbits` or Symbols. I don't think we need to commit to this extra restriction. `B` can be any function that returns a `Bijector`.

Reply:
> Type parameters must be isbits or Symbols.
So this is probably the issue I mention in things to discuss; is there somewhere where I can read about _when_ something is `isbits`?:)

And your suggestion seems like a good solution; I'll do that :+1:

Reply:
Should we just combine `B` and `F` then? The only way we use these are `B(cl.θ(x_2))`, so we could just instead remove the `B` field and let `F` be `B ∘ cl.θ`.

Reply:
A couple of reasons why I actually prefer the current approach vs. the two we've touched upon here:
- With `B` and `F` separate, we can dispatch differently on the coupling law `B` returning a `Bijector` and the parameter map `F`. E.g. define an `Affine(W, b) = Shift(b) ∘ Scale(W)` and then define `(cl::CouplingLayer{typeof(Affine)})(x)`. In _most_ cases this functionality won't be used, but at least with the current approach we have the option of doing this.
- With `B` in the type, we can do `CouplingLayer(Shift, M, F)` and things are type-stable. If we add it as a _field_, `CouplingLayer(Shift, mask, F)` results in a `UnionAll` type of the `B` field and thus we lose type-stability. As a result, to guarantee type-stability, we'd have combine `B` and `F` as mentioned above. But this means that we cannot dispatch differently on the coupling law and parameter map used, etc.

Reply:
Note that the type of `B` can be a type parameter. I don't see how this will harm type stability compared to directly using `B` itself as a type parameter.